src/cli/cli_coap.cpp

```diff
@@ -682,7 +685,24 @@ otError Coap::ProcessRequest(Arg aArgs[], otCoapCode aCoapCode)
     }
 #endif
 
-    SuccessOrExit(error = otCoapMessageAppendUriPathOptions(message, coapUri));
+    uriQueryStartPtr = StringFind(coapUri, '?');
+
+    if (uriQueryStartPtr == nullptr)
+    {
+        // "?" doesn't present in URI --> contains only URI path parts
+        SuccessOrExit(error = otCoapMessageAppendUriPathOptions(message, coapUri));
+    }
+    else
+    {
+        // "?" presents in URI --> contains URI path AND URI query parts
+        strncpy(coapUriQuery, uriQueryStartPtr + 1,
+                strlen(coapUri) - static_cast<size_t>(uriQueryStartPtr + 1 - coapUri));
+        memset(const_cast<char *>(uriQueryStartPtr), '\0',
+               strlen(coapUri) - static_cast<size_t>(uriQueryStartPtr - coapUri));
+
+        SuccessOrExit(error = otCoapMessageAppendUriPathOptions(message, coapUri));
+        SuccessOrExit(error = otCoapMessageAppendUriQueryOptions(message, coapUriQuery));
```

Comment:
Can consider reusing the same `coapUri` buffer for both strings (we won't need to copy into a new `coapUriQuery` buffer?) I think we just replace the `?` char with `\0` and it should work:

```suggestion
        // "?" presents in URI --> contains URI path AND URI query parts
        *uriQueryStartPtr++ = '\0';
        SuccessOrExit(error = otCoapMessageAppendUriPathOptions(message, coapUri));
        SuccessOrExit(error = otCoapMessageAppendUriQueryOptions(message, uriQueryStartPtr));
```